analysis/complex.lean

```diff
@@ -0,0 +1,342 @@
+@[simp] lemma I_re : complex.I.re=0 := rfl
+@[simp] lemma I_im : complex.I.im=1 := rfl
+@[simp] lemma conj_re (z : complex) : (conjugate z).re = z.re := rfl
+@[simp] lemma conj_im (z : complex) : (conjugate z).im = -z.im := rfl
+@[simp] lemma add_re (z w: complex) : (z+w).re=z.re+w.re := rfl
+@[simp] lemma add_im (z w: complex) : (z+w).im=z.im+w.im := rfl
+@[simp] lemma neg_re (z: complex) : (-z).re=-z.re := rfl
+@[simp] lemma neg_im (z: complex) : (-z).im=-z.im := rfl
+
+-- one size fits all tactic 
+
+meta def crunch : tactic unit := do
+`[intros],
+`[rw [eq_iff_re_eq_and_im_eq]],
+`[split;simp [add_mul,mul_add,mul_assoc]]
+
+meta def crunch2 : tactic unit := do 
+`[simp [add_mul, mul_add, eq_iff_re_eq_and_im_eq] {contextual:=tt}]
+
+
+meta def crunch3 : tactic unit := do 
+`[simp [add_mul, mul_add, mul_comm, mul_assoc, eq_iff_re_eq_and_im_eq] {contextual:=tt}]
+
+instance : add_comm_group complex :=
+{ add_comm_group .
+  zero         := 0,
+  add          := (+),
+  neg          := has_neg.neg,
+  zero_add     := by crunch,
+  add_zero     := by crunch,
+  add_comm     := by crunch,
+  add_assoc    := by crunch,
+  add_left_neg := by crunch
+}
+
+@[simp] lemma sub_re (z w : complex) : (z-w).re=z.re-w.re := rfl
+@[simp] lemma sub_im (z w : complex) : (z-w).im=z.im-w.im := rfl
+@[simp] lemma mul_re (z w: complex) : (z*w).re=z.re*w.re-z.im*w.im := rfl
+@[simp] lemma mul_im (z w: complex) : (z*w).im=z.re*w.im+z.im*w.re := rfl
+
+
+lemma norm_squared_pos_of_nonzero (z : complex) (H : z ≠ 0) : norm_squared z > 0 :=
+begin -- far more painful than it should be but I need it for inverses
+  suffices : z.re ≠ 0 ∨ z.im ≠ 0,
+  { apply lt_of_le_of_ne,
+    { exact add_nonneg (mul_self_nonneg _) (mul_self_nonneg _) },
+    intro H2,
+    cases this with Hre Him,
+    { exact Hre (eq_zero_of_mul_self_add_mul_self_eq_zero (eq.symm H2)) },
+    unfold norm_squared at H2,rw [add_comm] at H2,
+    exact Him (eq_zero_of_mul_self_add_mul_self_eq_zero (eq.symm H2)) },
+  have : ¬ (z.re = 0 ∧ z.im = 0),
+  { intro H2,
+    exact H (eq_of_re_eq_and_im_eq z 0 H2.left H2.right) },
+  by_cases z0 : (z.re = 0),-- with Hre_eq,-- Hre_ne,
+  { right,
+    intro H2,
+    apply this,
+    exact ⟨z0,H2⟩ },
+  { left,assumption }
+end
+
+lemma of_real_injective : function.injective of_real :=
+begin
+intros x₁ x₂ H,
+exact congr_arg complex.re H,
+end
+
+lemma of_real_zero : of_real 0 = (0:complex) := rfl
+lemma of_real_one : of_real 1 = (1:complex) := rfl
+lemma of_real_eq_coe (r : real) : of_real r = ↑r := rfl
+
+lemma of_real_neg (r : real) : -(r:complex) = ((-r:ℝ):complex) := by crunch
+
+lemma of_real_add (r s: real) : (r:complex) + (s:complex) = ((r+s):complex) := by crunch
+
+lemma of_real_sub (r s:real) : (r:complex) - (s:complex) = ((r-s):complex) := by crunch
+
+lemma of_real_mul (r s:real) : (r:complex) * (s:complex) = ((r*s):complex) := by crunch
+
+lemma of_real_inv (r:real) : (r:complex)⁻¹ = ((r⁻¹):complex) :=
+begin
```

Comment:
`of_real_add`, `of_real_sub`, `of_real_mul` and `of_real_inv` are a `rfl` lemmas (i.e. they can be proved by just `rfl`).

Reply:
Also: indentation after `begin`